DDS-800_recursive_deletes

dmlond opened the pull request.

dmlond committed:
- added #delete_children which deletes all child files, and creates FolderDeletionJobs out of each
  child folder
- added #manage_children which checks if is_updated has been changed to true,
  yields, then calls delete_children if is_updated has been changed to true
- calls #manage_children around_update

dmlond committed:
FolderDeletionJob updates a Folder to is_updated: true, which will
cascade through #manage_children

dmlond committed:
refactored manage_children and delete_children expectations into a shared example
'child minder'

FolderSpec expects Folder to behave like a child_minder
refactored some let variables to remove unused variables

dmlond committed:
ProjectSpec expects Project to behave like a child_minder
Project meets expectation

dmlond committed:
ensures that deleting a folder uses FolderDeletionjob appropriately

dmlond committed:
ProjectsApiSpec expects ProjectSpec to manage children when the project is deleted
and tests this in an inline and queued ActiveJob context
ProjectsApi meets expectation

dmlond committed:
refactored test for running a job into a rake shared example 'a queued job worker'
workers_rake_spec expects workers.rake to support
 - workers:initialize_project_storage:run
 - workers:delete_folders:run
workers.rake supports expected jobs
ProcFile defines process types:
  - web
  - project_storage_init
  - folder_deletion

dmlond committed:
shared_example logically deleted resource uses update_column instead of update_attribute
so that validations and callbacks are skipped. This makes it possible to delete a folder
without calling manage children

folders_api_spec no longer needs to run and stop the FolderDeletionJob before testing
logically deleted resources other than a call to DELETE a folder

dmann reviewed (changes requested):
Change FolderDeletionJob to ChildDeletionJob and move perform later to manage_children. Pass in serialized parent object instead of folder_id.

dmlond committed:
Refactored FolderDeletionJob to ChildDeletionJob

ChildMinder module implements manage_children
and delete_children

ModelSharedExamples implements
  - ChildMinder shared example which expects model to
    - include ChildMinder
    - have a callback to manage_children around update
  - ChildDeletionJob shared example which expects parent,
    child_file, and child_folder, and asserts that the
    ChildDeletionJob deletes all the child_files and child_folders

ProjectSpec expects Project to be a ChildMinder
FolderSpec expects Folder to be a ChildMinder

ChildDeletionJobSpec expects ChildDeletionJob to work with a project
and a folder as parent

dmlond committed:
name has random component to prevent periodic test failures due to 'name already taken' and limited
Faker dictionary diversity

dmlond committed:
works with new ChildDeletionJob expectations for deleting projects with root children

dmlond committed:
changed expectations to meet new ChildDeletionJob usage.
No longer expects DELETE folder to delete children since folders
are a ChildMinder

dmlond commented:
I made the requested changes. Thanks for the suggestions.

dmlond committed:
fixes bug with generated upload name, which needs to be an Internet.slug

dmlond committed:
needs to run and stop the ChildDeletionJob because its shared example
 expects to be able to update the project to is_deleted: true, and get
 an audit, which means all the callbacks have to run, including manage_children.

dmann commented on spec/factories/uploads.rb:
```diff
@@ -1,7 +1,7 @@
   factory :upload do
     project
-    name { Faker::Internet.slug }
+    name { "#{Faker::Internet.slug}#{rand(10**3)}" }
```
For uniqueness in a factory attribute, use a sequence:

``
sequence(:name) { |n| "#{Faker::Internet.slug(nil, '_')}_#{n}" }
``

Using rand will still create dups from time to time.

dmann commented on app/jobs/child_deletion_job.rb:
```diff
@@ -0,0 +1,9 @@
+class ChildDeletionJob < ApplicationJob
+  queue_as self.name.underscore.gsub('_job','').to_sym
```
This is a little complex for what amounts to `queue_as :child_deletion`  :\

dmann reviewed (changes requested):
I really like what you've done with the ChildMinder module and the ChildDeletionJob. One minor change to the uploads factory to prevent possible test failures.

dmlond committed:
simpler queue_as strings
upload name uses a sequence to ensure duplicates do not occur

dmlond commented:
ready for review again

dmann commented on Procfile:
```diff
@@ -1,2 +1,3 @@
 web: bundle exec puma -C config/puma.rb
-project_storage_init: rake workers:initialize_project_storage:run
+project_storage_init: bundle exec rake workers:initialize_project_storage:run
+folder_deletion: bundle exec rake workers:delete_folders:run
```
This should probably be renamed to `child_deletion` and `delete_children`

dmann commented on lib/tasks/workers.rake:
```diff
@@ -10,4 +10,15 @@ namespace :workers do
+      silence_warnings do
+        Rails.application.eager_load! unless Rails.application.config.eager_load
+      end
+      workers = [ FolderDeletionJob.job_wrapper ]
```
I'm surprised that this isn't causing tests to fail, since FolderDeletionJob no longer exists...

dmlond replied:
good catch. These dont run in circle.

dmann commented on spec/lib/tasks/workers_rake_spec.rb:
```diff
@@ -6,21 +6,18 @@
+  describe 'workers:delete_folders:run' do
+    include_context "rake"
+    let(:task_name) { "workers:delete_folders:run" }
+    let(:expected_job_class) { FolderDeletionJob }
```
Are the rake tests running in circle? Because this really should be failing...

dmann reviewed (changes requested):
Sorry that I didn't catch these in my last review. Lesson learned, review all files changed and not just the most recent changes...  :(

dmlond committed:
missed rake worker task that used FolderDeletionJob
renamed to delete_children and uses ChildDeletionJob

renamed ProcFile process types
  - all job runners end in _job (so we can scale foo_job 2)
  - changed folder_deletion to child_deletion_job
  - runs delete_children intead of delete_folders

dmlond commented:
ready for review again

dmlond committed:
worker_rake_spec test always runs in rspec
no more TEST_WORKERS=1 env requirement

dmann reviewed (approved):
👍
